logstash-core/lib/logstash/plugins/registry.rb

```diff
@@ -196,41 +197,53 @@ def lookup(type, plugin_name, &block)
     # a plugin and will do a lookup on the namespace of the required class to find a matching
     # plugin with the appropriate type.
     def legacy_lookup(type, plugin_name)
-      begin
-        path = "logstash/#{type}s/#{plugin_name}"
+      klass = load_plugin_class(type, plugin_name)
 
-        klass = begin
-          namespace_lookup(type, plugin_name)
-        rescue UnknownPlugin => e
-          # Plugin not registered. Try to load it.
-          begin
-            require path
-            namespace_lookup(type, plugin_name)
-          rescue LoadError => e
-            logger.error("Tried to load a plugin's code, but failed.", :exception => e, :path => path, :type => type, :name => plugin_name)
-            raise
-          end
-        end
+      has_alias = @alias_registry.alias?(type.to_java, plugin_name)
+      if !klass && has_alias
+        resolved_plugin_name = @alias_registry.original_from_alias(type.to_java, plugin_name)
+        logger.debug("Loading #{type} plugin #{resolved_plugin_name} via its alias #{plugin_name}...")
+        klass = load_plugin_class(type, resolved_plugin_name)
+      end
 
-        plugin = lazy_add(type, plugin_name, klass)
-      rescue => e
-        logger.error("Problems loading a plugin with",
-                    :type => type,
-                    :name => plugin_name,
-                    :path => path,
-                    :error_message => e.message,
-                    :error_class => e.class,
-                    :error_backtrace => e.backtrace)
-
-        raise LoadError, "Problems loading the requested plugin named #{plugin_name} of type #{type}. Error: #{e.class} #{e.message}"
+      unless klass
+        logger.error("Unable to load plugin.", :type => type, :name => plugin_name)
+        raise LoadError, "Unable to load the requested plugin named #{plugin_name} of type #{type}. The plugin is not installed."
+      end
+
+      plugin = lazy_add(type, plugin_name, klass)
+
+      if has_alias
+        @registry[key_for(type, resolved_plugin_name)] = plugin
       end
```

Comment:
This bit is what causes us to register the concrete plugin by its concrete name to our registry when we look for the _alias_. We will have already registered the concrete plugin by its alias via `lazy_add` above.

I don't think it is needed, and in fact it is problematic in a couple of ways. 

1. When the alias is registered before the concrete plugin, we bypass the log message about the concrete class being added to our registry (because it is done by `lazy_add`, and this injects directly).
2. But when the concrete plugin is registered before the alias, this stomps on the existing registered entry for the concrete plugin with a `PluginSpecification` that encodes the _alias_'s shape. This is not good.

I think we can skip it entirely.

```suggestion
```

If we do chose to add the concrete plugin to the registry when looked up via the alias, we should use `lazy_add`, and add the `klass` itself instead of reusing the (PluginSpecification) `plugin`:

```suggestion
      if has_alias
        lazy_add(type, resolved_plugin_name, klass)
      end
```

This unlocks a third option, which would be to avoid holding an `has_alias` and simply do the registration in the alias resolution block. I think I prefer this the most because it reduces the surface area of how aliases are intertwined with other lookups in this method.

Reply:
You are right and this code also generates a runtime error when a real plugin with a name corresponding to the alias is released.
In this case the `if !klass && has_alias` is false because the `klass` for the plugin exists, but `has_alias` is true, this makes to skip the assignment of `resolved_plugin_name` so it's undefined, and when it reaches
```ruby
if has_alias
  @registry[key_for(type, resolved_plugin_name)] = plugin
end
```
it generates an error